diff --git a/src/corelib/Qt6CoreMacros.cmake b/src/corelib/Qt6CoreMacros.cmake
index bfac685..f40138e 100644
--- a/src/corelib/Qt6CoreMacros.cmake
+++ b/src/corelib/Qt6CoreMacros.cmake
@@ -3243,7 +3243,7 @@ function(_qt_internal_setup_deploy_support)
         endif()
     endforeach()

-    set(qtpaths_prefix "${QT6_INSTALL_PREFIX}/${QT6_INSTALL_BINS}")
+    set(qtpaths_prefix "${QT6_INSTALL_PREFIX}/tools/Qt6/bin")

     set(candidate_paths "")
     foreach(qtpaths_name_candidate IN LISTS qtpaths_name_candidates)
