diff --git a/src/plugins/position/corelocation/qgeopositioninfosource_cl.mm b/src/plugins/position/corelocation/qgeopositioninfosource_cl.mm
index 95d51a86..7529df9d 100644
--- a/src/plugins/position/corelocation/qgeopositioninfosource_cl.mm
+++ b/src/plugins/position/corelocation/qgeopositioninfosource_cl.mm
@@ -51,7 +51,7 @@
     // Construct position info from location data
     QGeoPositionInfo location(QGeoCoordinate(newLocation.coordinate.latitude,
                                              newLocation.coordinate.longitude,
-                                             newLocation.altitude),
+                                             newLocation.ellipsoidalAltitude),
                                              timeStamp);
     if (newLocation.horizontalAccuracy >= 0)
         location.setAttribute(QGeoPositionInfo::HorizontalAccuracy, newLocation.horizontalAccuracy);
